core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -132,203 +126,84 @@ protected boolean resolveIndex() {
         return new WriteResult<>(response, location);
     }
 
-    private Translog.Location handleItem(IndexMetaData metaData, BulkShardRequest request, IndexShard indexShard, long[] preVersions, VersionType[] preVersionTypes, Translog.Location location, int requestIndex, BulkItemRequest item) {
-        if (item.request() instanceof IndexRequest) {
-            location = index(metaData, request, indexShard, preVersions, preVersionTypes, location, requestIndex, item);
-        } else if (item.request() instanceof DeleteRequest) {
-            location = delete(request, indexShard, preVersions, preVersionTypes, location, requestIndex, item);
-        } else if (item.request() instanceof UpdateRequest) {
-            Tuple<Translog.Location, BulkItemRequest> tuple = update(metaData, request, indexShard, preVersions, preVersionTypes, location, requestIndex, item);
-            location = tuple.v1();
-            item = tuple.v2();
-        } else {
-            throw new IllegalStateException("Unexpected index operation: " + item.request());
-        }
-
-        assert item.getPrimaryResponse() != null;
-        assert preVersionTypes[requestIndex] != null;
-        return location;
-    }
-
-    private Translog.Location index(IndexMetaData metaData, BulkShardRequest request, IndexShard indexShard, long[] preVersions, VersionType[] preVersionTypes, Translog.Location location, int requestIndex, BulkItemRequest item) {
-        IndexRequest indexRequest = (IndexRequest) item.request();
-        preVersions[requestIndex] = indexRequest.version();
-        preVersionTypes[requestIndex] = indexRequest.versionType();
+    /** Executes bulk item requests and handles request execution exceptions */
+    private Translog.Location executeBulkItemRequest(IndexMetaData metaData, IndexShard indexShard,
+                                                     BulkShardRequest request,
+                                                     long[] preVersions, VersionType[] preVersionTypes,
+                                                     Translog.Location location, int requestIndex) {
+        preVersions[requestIndex] = request.items()[requestIndex].request().version();
+        preVersionTypes[requestIndex] = request.items()[requestIndex].request().versionType();
+        DocumentRequest.OpType opType = request.items()[requestIndex].request().opType();
         try {
-            WriteResult<IndexResponse> result = shardIndexOperation(request, indexRequest, metaData, indexShard, true);
-            location = locationToSync(location, result.getLocation());
+            WriteResult<? extends DocWriteResponse> writeResult = innerExecuteBulkItemRequest(metaData, indexShard,
+                request, requestIndex);
+            if (writeResult.getResponse().getResult() != DocWriteResponse.Result.NOOP) {
```

Comment:
I wanted to be more strict as only a noop result can be valid with a null location. changed it to be an assert instead.